core/systems/assignment/template.ex

```diff
@@ -23,33 +23,188 @@ defprotocol Systems.Assignment.Template do
 end
 
 defmodule Systems.Assignment.Template.Flags do
-  defmacro __using__(flags) do
+  @moduledoc """
+  A macro for creating flag structs that control feature availability.
+
+  This module provides a macro that generates flag structs with opt-in behavior:
+  - By default, all flags are disabled (false)
+  - Use `opt_in: [list_of_flags]` to enable specific flags
+  - `opt_out` is deprecated but still supported with warnings
+
+  ## Usage
+
+      defmodule MyFlags do
+        use Systems.Assignment.Template.Flags, [:feature_a, :feature_b, :feature_c]
+      end
+
+      # Create with all flags disabled (default)
+      flags = MyFlags.new()
+
+      # Create with specific flags enabled
+      flags = MyFlags.new(opt_in: [:feature_a, :feature_c])
+  """
+
+  defmacro __using__(list_of_available_flags) do
     quote do
+      @available_flags unquote(list_of_available_flags)
+
+      @typedoc """
+      A struct containing boolean flags for controlling feature availability.
+      Each flag can be true (enabled) or false (disabled).
+      """
       @type t :: %__MODULE__{
               unquote_splicing(
-                for flag <- flags do
-                  {flag, {:boolean, [], []}}
+                for flag_name <- list_of_available_flags do
+                  {flag_name, {:boolean, [], []}}
                 end
               )
             }
 
-      defstruct unquote(flags)
+      defstruct unquote(list_of_available_flags)
+
+      @doc """
+      Returns the complete list of available flags for this module.
+      Useful for introspection and validation.
+      """
+      @spec flags() :: [atom()]
+      def flags, do: @available_flags
+
+      @doc """
+      Creates a new flags struct with specified options.
+
+      ## Options
+
+      * `:opt_in` - List of flags to enable (recommended)
+      * `:opt_out` - List of flags to disable (deprecated)
+
+      ## Examples
+
+          iex> #{inspect(__MODULE__)}.new(opt_in: [:flag1, :flag2])
+          %#{inspect(__MODULE__)}{flag1: true, flag2: true, ...}
+
+          iex> #{inspect(__MODULE__)}.new()
+          %#{inspect(__MODULE__)}{flag1: false, flag2: false, ...}
+      """
+      @spec new(keyword()) :: t()
+      def new(options_list \\ []) do
+        # Check for deprecated opt_out usage and warn user
+        _warn_if_using_deprecated_opt_out(options_list)
+        # Determine which approach to use based on provided options
+        cond do
+          _has_opt_in_option?(options_list) ->
+            _create_struct_with_opt_in_flags(options_list, unquote(list_of_available_flags))
+
+          _has_opt_out_option?(options_list) ->
+            _create_struct_with_opt_out_flags(options_list, unquote(list_of_available_flags))
+
+          true ->
+            _create_struct_with_all_flags_disabled(unquote(list_of_available_flags))
+        end
+      end
+
+      # Private helper functions for better readability
+
+      defp _warn_if_using_deprecated_opt_out(options_list) do
```

Comment:
Reconsider to start private function name with an '_'.

Reply:
Bedoel je hiermee dat je het liever niet hebt, de `_`?